src/streaming/array.writer.cpp

```diff
@@ -0,0 +1,422 @@
+
+    downsampled_config.dtype = config.dtype;
+
+    // copy the Blosc compression parameters
+    downsampled_config.compression_params = config.compression_params;
+
+    // can we downsample downsampled_config?
+    for (auto i = 0; i < config.dimensions->ndims(); ++i) {
+        // downsampling made the chunk size strictly smaller
+        const auto& dim = config.dimensions->at(i);
+        const auto& downsampled_dim = downsampled_config.dimensions->at(i);
+
+        if (dim.chunk_size_px > downsampled_dim.chunk_size_px) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/// Writer
+zarr::ArrayWriter::ArrayWriter(
+  const ArrayWriterConfig& config,
+  std::shared_ptr<ThreadPool> thread_pool,
+  std::shared_ptr<S3ConnectionPool> s3_connection_pool)
+  : config_{ config }
+  , thread_pool_{ thread_pool }
+  , s3_connection_pool_{ s3_connection_pool }
+  , bytes_to_flush_{ 0 }
+  , frames_written_{ 0 }
+  , append_chunk_index_{ 0 }
+  , is_finalizing_{ false }
+{
+}
+
+size_t
+zarr::ArrayWriter::write_frame(const uint8_t* data, size_t nbytes)
+{
+    const size_t nbytes_frame =
+      bytes_of_frame(*config_.dimensions, config_.dtype);
+
+    if (nbytes_frame != nbytes) {
+        LOG_WARNING("Frame size mismatch: expected %zu, got %zu. Skipping",
+                    nbytes_frame,
+                    nbytes);
+        return 0;
+    }
+
+    if (chunk_buffers_.empty()) {
+        make_buffers_();
+    }
+
+    // split the incoming frame into tiles and write_frame them to the chunk
+    // buffers
+    const auto bytes_written = write_frame_to_chunks_(data, nbytes);
+    EXPECT(bytes_written == nbytes, "Failed to write_frame frame to chunks");
+
+    LOG_DEBUG("Wrote %zu bytes of frame %zu", bytes_written, frames_written_);
+    bytes_to_flush_ += bytes_written;
+    ++frames_written_;
+
+    if (should_flush_()) {
+        flush_();
+    }
+
+    return bytes_written;
+}
+
+bool
+zarr::ArrayWriter::make_data_sinks_()
+{
+    std::string data_root;
+    std::function<size_t(const ZarrDimension&)> parts_along_dimension;
+    switch (version_()) {
+        case ZarrVersion_2:
+            parts_along_dimension = chunks_along_dimension;
+            data_root = config_.store_path + "/" +
+                        std::to_string(config_.level_of_detail) + "/" +
+                        std::to_string(append_chunk_index_);
+            break;
+        case ZarrVersion_3:
+            parts_along_dimension = shards_along_dimension;
+            data_root = config_.store_path + "/data/root/" +
+                        std::to_string(config_.level_of_detail) + "/c" +
+                        std::to_string(append_chunk_index_);
+            break;
+        default:
+            LOG_ERROR("Unsupported Zarr version");
+            return false;
+    }
+
+    SinkCreator creator(thread_pool_, s3_connection_pool_);
+
+    if (config_.bucket_name && !creator.make_data_sinks(*config_.bucket_name,
```

Comment:
Are you using `bucket_name` to determine if an S3 sink should be created? This approach can lead to issues and doesn't enhance readability. I suggest creating a computed property in the configuration object—something like `is_s3_sink`—that encapsulates this logic. This approach allows you to modify the logic in one central location if needed, while also clearly conveying your intent.